README.md
<img alt="Logo for fovea" src="./documentation/asset/fovea_icon.png" height="200"></img><br>
<a href="https://www.npmjs.com/package/@fovea/core"><img alt="NPM Version" src="https://badge.fury.io/js/%40fovea%2Fcore.svg" height="20"></img></a>
<a href="https://github.com/fovea-org/fovea/graphs/contributors"><img alt="Contributors" src="https://img.shields.io/github/contributors/fovea-org%2Ffovea.svg" height="20"></img></a>
<a href="https://opensource.org/licenses/MIT"><img alt="MIT License" src="https://img.shields.io/badge/License-MIT-yellow.svg" height="20"></img></a>
<a href="https://www.patreon.com/bePatron?u=11315442"><img alt="Support on Patreon" src="https://c5.patreon.com/external/logo/become_a_patron_button@2x.png" height="20"></img></a>

## Introduction

Fovea is a compiler for generating tiny and powerful Web Components, to use in building Web applications.
By shifting computations to compile-time, rather than run-time, Fovea enables you to do way more with way less overhead.
And, because components compiled with Fovea are just Web Components, they work in any framework you may choose - or with no framework at all.
Fovea also makes it super easy to build powerful Progressive Web Apps with the [`@fovea/cli`](https://github.com/fovea-org/fovea/blob/master/packages/cli/README.md).
There's so much we want to say, but let's take it bit by bit:

## Install

- [Build Progressive Web Apps with Fovea] by leveraging the powerful `@fovea/cli`.
- [Build reusable Web Components with Fovea] with the `@fovea/compiler` and `@fovea/rollup-plugin-fovea`.

## Examples

A Fovea component looks like a Custom Element - because it _is_ a Custom Element! And, your templates and styles are completely encapsulated within the Shadow DOM.
Here's a very simple example of a Fovea component:

```typescript
@templateSrc("./my-component.html")
@styleSrc(["./my-component.scss"])
class MyComponent extends HTMLElement {}
```

The decorators are just there to help Fovea understand what you want to achieve. They are gone in the compiled output!
Fovea enables you to write your styles in `.css` (or `.scss`) and your template in `.html`, - but it doesn't force you to!
If you're a fan of [React], [lit-html] or [Stencil], you can also declare both directly from the component with `template` and/or `styles` accessors, methods or property declarations.
Because Fovea is compiled, it doesn't matter how you provide your template and styles. Fovea will figure it out:

```typescript
class MyComponent extends HTMLElement {
  // You can also provide the template from a get accessor or from a method of the same name
  // In this example, it is provided from a property.
  template = `<h1>This is my component!</h1>`;
}
```

By design, Fovea let's you write as little as possible, and fills out the gaps for you as required. In this example:

- The `MyComponent` Custom Element is automatically registered with the selector `my-component`
- A Shadow Root is automatically attached to the element
- The template is compiled into a sequence of blazing fast, secure instructions and attached the Shadow root
- The styles are run through PostCSS and compiled with SASS if applicable, and attached to the Shadow root
  But you can manually invoke `customElements.define(...)` if you want, or provide a different selector name. Fovea won't get in your way.

Here's another example of a component that does a little more:

<!-- prettier-ignore -->
```typescript
@selector("clickey-thingey")
class MyComponent extends HTMLElement {
  @prop @setOnHost clickAmount: number = 0;

  template = `
		<button on-click="${this.clickAmount++}">
		I've been clicked ${this.clickAmount} ${this.clickAmount === 1 ? "time" : "times"}!
		</button>`;

  @onChange("clickAmount")
  onClickAmountChanged() {
    if (this.clickAmount >= 10) {
      console.log("You should go play Cookie clicker instead!");
    }
  }
}
```

This example is a bit more involved, but showcase some of the strongest parts of Fovea:

- You can declare `@prop`s which are the properties you'd like to use within your templates and styles. Here, the `prop` `clickAmount` of type `number` is declared (yes, the data type is actually used for coercion!)
- You can use complex Expressions inside your templates for everything that goes into the `${...}` parts. We call them _Live Expressions_, because they react immediately to changes to the `prop`s that are used within them. You can use _all_ Javascript language features within those.
- You can automatically sync your `props` with host attributes. Fovea takes care of mapping the names between them. In this example, the `prop` `clickAmount` will be set on the host element as the attribute `click-amount` and update each time the value changes. It also works the other way around - when the attributes change, the `prop` is updated - and coerced into the data type that it is declared with.
- You can react to changes to your props with the `@onChange` decorator.

Best of all, you don't pay any cost for the features that you don't use. For example, if you never use the `@onChange` decorator, Fovea won't ship the related logic.
Naturally, Fovea components can do **so** much more, so instead we'll refer you to the [full examples section].

## Contributing

Do you want to contribute? Awesome! Please follow [these recommendations](./CONTRIBUTING.md).

## Maintainers

- <a href="https://github.com/wessberg"><img alt="Frederik Wessberg" src="https://avatars2.githubusercontent.com/u/20454213?s=460&v=4" height="11"></img></a> [Frederik Wessberg](https://github.com/wessberg): _Lead Developer_

## Backers 🏅

[Become a backer](https://www.patreon.com/bePatron?u=11315442) and get your name, logo, and link to your site listed here.

## License 📄

MIT © [Frederik Wessberg](https://github.com/wessberg)

## FAQ

#### Another UI library? Are you serious?

Yes, we're quite serious. Remember, you get Web Components out! Fovea is build on standards - we're not trying to build a new ecosystem around yet another framework. Everything
that Fovea enables you to produce is compatible with every other framework out there, including the one you might be using right now.
Fovea wants to place itself as something in-between [Stencil] and [Polymer]. Stencil is a compiler as well, but it is closer to `React` in technology (JSX, Virtual DOM) and philosophy (Functional, Reactive) than Fovea is.
Polymer is built on the same core principles, but it doesn't need a compilation step and comes with the benefits (no tooling required) and constraints (greater runtime cost) that result from it.

#### Is there a way to use Fovea without the compile-step?

**No**. The reason why Fovea is so fast and so small is because it knows everything there is to know about your code before any browser ever gets involved. That enables Fovea to cool things such as convert your templates into assembly-like instructions, as well as use the data types of your props for things like coercion.
Fovea is, and always will be, a compile-time tool.

[react]: https://reactjs.org/
[stencil]: https://stenciljs.com/
[polymer]: https://www.polymer-project.org/
[lit-html]: https://polymer.github.io/lit-html/


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.137"></a>
## [1.0.137](https://github.com/fovea-org/fovea/compare/v1.0.136...v1.0.137) (2018-09-30)

**Note:** Version bump only for package fovea





<a name="1.0.136"></a>
## [1.0.136](https://github.com/fovea-org/fovea/compare/v1.0.135...v1.0.136) (2018-09-30)

**Note:** Version bump only for package fovea





<a name="1.0.135"></a>
## [1.0.135](https://github.com/fovea-org/fovea/compare/v1.0.134...v1.0.135) (2018-09-30)


### Features

* [[@fovea](https://github.com/fovea)/router]: Routes now support a  property for routes that has children. When no child is matched by the URL, the child matched by the  property will be used. This fixes [#8](https://github.com/fovea-org/fovea/issues/8) ([ab259f7](https://github.com/fovea-org/fovea/commit/ab259f7))





<a name="1.0.134"></a>
## [1.0.134](https://github.com/fovea-org/fovea/compare/v1.0.133...v1.0.134) (2018-09-29)


### Features

* Added support for forced attributes. Fixes [#30](https://github.com/fovea-org/fovea/issues/30) ([60d4f4c](https://github.com/fovea-org/fovea/commit/60d4f4c))
* Improved minification for production builds ([84c3995](https://github.com/fovea-org/fovea/commit/84c3995))
* Made advanced treeshaking customization customizable from a new 'optimization' property of outputs declared in fovea-cli.config files ([f3db25e](https://github.com/fovea-org/fovea/commit/f3db25e))
* Made the fovea-cli.config format allow for even more customization of the build pipeline, including compression, minification, comment preserving/removal logic, advanced treeshaking, intro/outro/banner/footer appending/prepending, as well as sourcemap handling. ([6d5ff05](https://github.com/fovea-org/fovea/commit/6d5ff05))





<a name="1.0.133"></a>
## [1.0.133](https://github.com/fovea-org/fovea/compare/v1.0.132...v1.0.133) (2018-09-27)


### Features

* Added major scheduling overhaul for performance gains. Rendering is now asynchronous. More details will arrive in a detailed post ([7371991](https://github.com/fovea-org/fovea/commit/7371991))
* Optimizations on allocated heap memory ([2768bd1](https://github.com/fovea-org/fovea/commit/2768bd1))
* Removed external dependency from [@fovea](https://github.com/fovea)/router for smaller size. ([1cb3d96](https://github.com/fovea-org/fovea/commit/1cb3d96))





<a name="1.0.132"></a>
## [1.0.132](https://github.com/fovea-org/fovea/compare/v1.0.131...v1.0.132) (2018-09-20)


### Bug Fixes

* Fixed a few remaining issues with taking variables inside [@fovea](https://github.com/fovea)/styles ([9da4192](https://github.com/fovea-org/fovea/commit/9da4192))
* Fixed an issue where the CLI wouldn't generate a package.json that depended on [@fovea](https://github.com/fovea)/common. This would break hoisted dependencies in hoisted monorepos ([f4d21a0](https://github.com/fovea-org/fovea/commit/f4d21a0))
* Fixed an issue where the CLI wouldn't generate a package.json that depended on [@fovea](https://github.com/fovea)/lib ([f4d7b90](https://github.com/fovea-org/fovea/commit/f4d7b90))
* Fixed an issue with radio buttons in [@fovea](https://github.com/fovea)/material. Improved docs ([0aba61a](https://github.com/fovea-org/fovea/commit/0aba61a))
* Fixed an issue with taking variables which pointed to CSS Custom Properties with default values. The new behavior makes sure to actually use the default value in JSON output. ([fb87897](https://github.com/fovea-org/fovea/commit/fb87897))
* Fixed performance issues with debouncing and lazily evaluating certain things in [@fovea](https://github.com/fovea)/material ([10bd7e5](https://github.com/fovea-org/fovea/commit/10bd7e5))


### Features

* .[sass|scss] files can now [@import](https://github.com/import) from node_modules with the '~' prefix using the node resolution algorithm. ([10fa8a2](https://github.com/fovea-org/fovea/commit/10fa8a2))
* Abstract classes that directly or indirectly derives from HTMLElement will no longer auto-generate calls to customElements.define or customAttributes.define since these are not constructable and meant for subclassing. Fixes [#27](https://github.com/fovea-org/fovea/issues/27) ([f3b1168](https://github.com/fovea-org/fovea/commit/f3b1168))
* Added better runtime logging for when an already declared Custom Element or Custom Attribute is attempted to be defined. Fixes [#24](https://github.com/fovea-org/fovea/issues/24) ([be25025](https://github.com/fovea-org/fovea/commit/be25025))
* Made icon-button-components compatible with forms with the 'button' role ([053ce34](https://github.com/fovea-org/fovea/commit/053ce34))
* Made the base font size themeable in [@fovea](https://github.com/fovea)/material ([3ce8e9f](https://github.com/fovea-org/fovea/commit/3ce8e9f))
* Made transition durations and timing functions themeable in [@fovea](https://github.com/fovea)/material ([b32ed55](https://github.com/fovea-org/fovea/commit/b32ed55))


### BREAKING CHANGES

* The button Custom Attribute in @fovea/material is now a proper custom element to enhance interoperability.





<a name="1.0.131"></a>
## [1.0.131](https://github.com/fovea-org/fovea/compare/v1.0.130...v1.0.131) (2018-09-12)

**Note:** Version bump only for package fovea





<a name="1.0.130"></a>
## [1.0.130](https://github.com/fovea-org/fovea/compare/v1.0.129...v1.0.130) (2018-09-11)


### Bug Fixes

* Fixed an issue with taking [css|scss] variables from theme styles. ([848b25b](https://github.com/fovea-org/fovea/commit/848b25b))





<a name="1.0.129"></a>
## [1.0.129](https://github.com/fovea-org/fovea/compare/v1.0.128...v1.0.129) (2018-09-10)


### Bug Fixes

* Fixed a bug where element names inside SVG markup would be forced into dash-case, even though they may be case-sensitive ([6e934c7](https://github.com/fovea-org/fovea/commit/6e934c7))

shared/rollup.browser.config.js
import baseConfig from "./rollup.config";
import nodeResolve from "rollup-plugin-node-resolve";
import commonjs from "rollup-plugin-commonjs";
import minify from "rollup-plugin-babel-minify";
import {FORMAT_FLAT_MODULE, FORMAT_IMMEDIATELY_INVOKED_FUNCTION, FORMAT_MODULE, formatOutputPath} from "./shared-package.json.js";

const browserConfig = {
	...baseConfig,
	plugins: [
		commonjs(),
		nodeResolve(),
		...baseConfig.plugins,
		...(process.env.NODE_ENV !== "production" ? [] : [
			minify({
				comments: false
			})
		])
	]
};

export default (name) => [
	{
		isFlat: false,
		config: {
			...browserConfig,
			external: [
				...browserConfig.external,
				"@fovea/lib"
			]
		}
	},
	{
		isFlat: true,
		config: {
			...browserConfig,
			output: [
				{
					file: formatOutputPath(FORMAT_FLAT_MODULE),
					format: FORMAT_MODULE,
					sourcemap: true
				},
				{
					file: formatOutputPath(FORMAT_IMMEDIATELY_INVOKED_FUNCTION),
					format: FORMAT_IMMEDIATELY_INVOKED_FUNCTION,
					name,
					sourcemap: true
				}
			],
			external: [
				...browserConfig.external
			]
		}
	}
];

shared/shared-package.json.js
import packageJSON from "../package.json";
export const FORMAT_NODE = "cjs";
export const FORMAT_MODULE = "esm";
export const FORMAT_IMMEDIATELY_INVOKED_FUNCTION = "iife";
export const FORMAT_FLAT_MODULE = "fesm";

export const formatOutputPath = (format, ext = ".js") =>  `./dist/${format}/index${ext}`;

/**
 * These fields are shared among all package.json files across the project
 */
export default {
	main: formatOutputPath(FORMAT_NODE),
	module: formatOutputPath(FORMAT_MODULE),
	browser: formatOutputPath(FORMAT_MODULE),
	types: formatOutputPath(FORMAT_MODULE, ".d.ts"),
	typings: formatOutputPath(FORMAT_MODULE, ".d.ts"),
	es2015: formatOutputPath(FORMAT_MODULE),
	files: [
		"dist/**/*.*",
		"bin/*"
	],
	repository: packageJSON.repository,
	bugs: packageJSON.bugs,
	contributors: packageJSON.contributors,
	engines: packageJSON.engines,
	license: packageJSON.license,
	scaffold: {
		patreonUserId: "11315442",
		contributorMeta: {
			"Frederik Wessberg": {
				imageUrl: "https://avatars2.githubusercontent.com/u/20454213?s=460&v=4",
				role: "Lead Developer",
				twitterHandle: "FredWessberg",
				isCocEnforcer: true
			}
		},
		backers: []
	}
};